pkgs/development/mobile/androidenv/ndk-bundle/default.nix

```diff
@@ -58,6 +59,14 @@ deployAndroidPackage rec {
     for progname in ndk-build; do
       wrapProgram "$(pwd)/$progname" --prefix PATH : "${runtime_paths}"
     done
+    # wrab executables in toolchains
+    find $out/libexec/android-sdk/ndk/${package.revision}/toolchains -type d -name bin | while read bindir; do
+      find $bindir -maxdepth 1 -type f -executable | while read progname; do
+        wrapProgram "$progname"\
+          --prefix PATH : "${runtime_paths}"\
+          --prefix LD_LIBRARY_PATH : ${lib_paths}
```

Comment:
I tried something similar to this approach when I was looking into #298285, but that broke `aapt` because `libc++.so` fom nixpkgs is linker script:
```shell
[nix-shell:~/dev/sandbox]$ cat shell.nix 
with (import <nixpkgs> ) {
  config = {
    android_sdk.accept_license = true;
    allowUnfree = true;
  };
};
let
    androidComposition = androidenv.composeAndroidPackages {
      platformVersions = [ "34" ];
      includeNDK = true;
      ndkVersion = "26.1.10909125";
      abiVersions = [ "x86_64" ];
  };
  androidSdk = androidComposition.androidsdk;
in
stdenv.mkDerivation rec {
  name = "test-env";

  buildInputs = [
    androidSdk
  ];

  ANDROID_SDK_ROOT = "${androidSdk}/libexec/android-sdk";
  ANDROID_NDK_ROOT = "${ANDROID_SDK_ROOT}/ndk-bundle";
  LD_LIBRARY_PATH = lib.makeLibraryPath ([ libcxx ]);
}
```
```shell
[nix-shell:~/dev/sandbox]$ $ANDROID_SDK_ROOT/build-tools/34.0.0/aapt
/nix/store/52h1027gg0mbwyaba79rgzl98hb7jxyv-androidsdk/libexec/android-sdk/build-tools/34.0.0/aapt: error while loading shared libraries: /nix/store/zihlra5wv0jahps2dq0w4krbwy32zy04-libcxx-16.0.6/lib/libc++.so: file too short
```
```shell
[nix-shell:~/dev/sandbox/]$ cat /nix/store/zihlra5wv0jahps2dq0w4krbwy32zy04-libcxx-16.0.6/lib/libc++.so
INPUT(libc++.so.1 -lc++abi)
```



Disclaimer: I'm not sure if this is entirely equivalent to the change in the PR and I haven't had a chance to test this patch.

Reply:
This is how I test these stuff: Of course I have a clone off `nixpkgs` that's checked out to the branch of this PR. The `shell.nix` file is the one that you gave me in the issue itself with a bit of modification:
```nix
with (import /PATH/TO/THE/nixpkgs ) {
  config = {
    android_sdk.accept_license = true;
    allowUnfree = true;
  };
};
let
    androidComposition = androidenv.composeAndroidPackages {
      platformVersions = [ "34" ];
      includeNDK = true;
      ndkVersion = "26.1.10909125";
      abiVersions = [ "x86_64" ];
  };
  androidSdk = androidComposition.androidsdk;
  androidSdkRoot = "${androidSdk}/libexec/android-sdk";
in
stdenv.mkDerivation {
  name = "test-env";

  buildInputs = [
    androidSdk
  ];

  ANDROID_SDK_ROOT = androidSdkRoot;
  ANDROID_NDK_ROOT = "${androidSdkRoot}/ndk-bundle";
}
```
Then the followings show my results:
```shell
$ nix-shell --run '$ANDROID_NDK_ROOT/toolchains/llvm/prebuilt/linux-x86_64/bin/clang -v'
Android (10552028, +pgo, +bolt, +lto, -mlgo, based on r487747d) clang version 17.0.2 (https://android.googlesource.com/toolchain/llvm-project d9f89f4d16663d5012e5c09495f3b30ece3d2362)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /nix/store/nvasmc0ymqbw20a0i204lgf2hv27pwzh-androidsdk/libexec/android-sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/bin
```
and 
```shell
$ nix-shell --run '$ANDROID_SDK_ROOT/build-tools/34.0.0/aapt'                                                         
Android Asset Packaging Tool                                                                                          
                                                                                                                      
Usage:                                                                                                                
 aapt l[ist] [-v] [-a] file.{zip,jar,apk}                                                                             
   List contents of Zip-compatible archive.
....
```
and
```shell
$ cat /nix/store/ij6lcyk0x1gj04nzx9m3ph73lv4f8x9y-libcxx-16.0.6/lib/libc++.so
INPUT(libc++.so.1 -lc++abi)
```
These looks okay to me. The reason that `aapt` didn't work for your workaround is simply because you have set `LD_LIBRARY_PATH` globally, which could break other packages, while in this PR by wrapping the specific executables in the toolchains directory with `LD_LIBRARY_PATH` we avoid side effects to other packages. Let me know what do you think.

Reply:
Thanks for the explanation @hadilq, it makes sense globally overriding `LD_LIBRARY_PATH` was the issue in my case. Sadly, I won't be time to test this patch until next week, so feel free to go ahead. Thanks again for fixing this!